In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotly.express as px

In [2]:
df_import= pd.read_csv('C:\\Users\\HP Envy\\Desktop\\Data analyst\\archive\\ds_salaries.csv')
df_import

Unnamed: 0  work_year experience_level employment_type  \
0             0       2020               MI              FT   
1             1       2020               SE              FT   
2             2       2020               SE              FT   
3             3       2020               MI              FT   
4             4       2020               SE              FT   
..          ...        ...              ...             ...   
602         602       2022               SE              FT   
603         603       2022               SE              FT   
604         604       2022               SE              FT   
605         605       2022               SE              FT   
606         606       2022               MI              FT   

                      job_title  salary salary_currency  salary_in_usd  \
0                Data Scientist   70000             EUR          79833   
1    Machine Learning Scientist  260000             USD         260000   
2             Big Data Engineer   85000             GBP         109024   
3          Product Data Analyst   20000             USD          20000   
4     Machine Learning Engineer  150000             USD         150000   
..                          ...     ...             ...            ...   
602               Data Engineer  154000             USD         154000   
603               Data Engineer  126000             USD         126000   
604                Data Analyst  129000             USD         129000   
605                Data Analyst  150000             USD         150000   
606                AI Scientist  200000             USD         200000   

    employee_residence  remote_ratio company_location company_size  
0                   DE             0               DE            L  
1                   JP             0               JP            S  
2                   GB            50               GB            M  
3                   HN             0               HN            S  
4                   US            50               US            L  
..                 ...           ...              ...          ...  
602                 US           100               US            M  
603                 US           100               US            M  
604                 US             0               US            M  
605                 US           100               US            M  
606                 IN           100               US            L  

[607 rows x 12 columns]

# Understanding the data

The dataframe has 11 columns (without unnamed): 
1. Work year: The year that data was collected 
2. Experience_level: The experience_level in the job, or you can understand that experience level is your level in companies, including:
* EN = Entry level
* MI = Mid level
* SE = Senior level
* EX = Executive level
3. Employment_type: Employment_type helps you to know whether the staff works full-time or part-time.... This includes:
* PT = Part-time
* FT = Full-time
* CT = Contract
* FL = Freelance
4. Job_title: Job_title is the name of your job
5. Salary: the money that staff is paid due to their work
6. Salary_currency: The currency that salary is paid. It can be paid in USD, EURO,...
7. SalaryinUSD: The salary is paid in USD (United State Dollar)
8. Employee_residence: Employee_residence is the nation that staff was working in
9. remote_ratio: this calculate the amount of work done remotely:
* 0 = no remote work (work onsite)
* 50 = Partially remote (half of work is done remotely, half of work is done onsite)
* 100 = fully remote
10. company_loacation: Determining the nation of your company
11. Company_size: the size of companies that staff has worked in, including:
* S = less than 50 employees (small)
* M = 50 to 250 employees (medium)
* L = more than 250 employees (large)

# IMPORTING LIBRARIES AND IMPORTING DATA

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import plotly.express as px

In [4]:
df_import= pd.read_csv('C:\\Users\\HP Envy\\Desktop\\Data analyst\\archive\\ds_salaries.csv')
df_import

Unnamed: 0  work_year experience_level employment_type  \
0             0       2020               MI              FT   
1             1       2020               SE              FT   
2             2       2020               SE              FT   
3             3       2020               MI              FT   
4             4       2020               SE              FT   
..          ...        ...              ...             ...   
602         602       2022               SE              FT   
603         603       2022               SE              FT   
604         604       2022               SE              FT   
605         605       2022               SE              FT   
606         606       2022               MI              FT   

                      job_title  salary salary_currency  salary_in_usd  \
0                Data Scientist   70000             EUR          79833   
1    Machine Learning Scientist  260000             USD         260000   
2             Big Data Engineer   85000             GBP         109024   
3          Product Data Analyst   20000             USD          20000   
4     Machine Learning Engineer  150000             USD         150000   
..                          ...     ...             ...            ...   
602               Data Engineer  154000             USD         154000   
603               Data Engineer  126000             USD         126000   
604                Data Analyst  129000             USD         129000   
605                Data Analyst  150000             USD         150000   
606                AI Scientist  200000             USD         200000   

    employee_residence  remote_ratio company_location company_size  
0                   DE             0               DE            L  
1                   JP             0               JP            S  
2                   GB            50               GB            M  
3                   HN             0               HN            S  
4                   US            50               US            L  
..                 ...           ...              ...          ...  
602                 US           100               US            M  
603                 US           100               US            M  
604                 US             0               US            M  
605                 US           100               US            M  
606                 IN           100               US            L  

[607 rows x 12 columns]

# UNDERSTANDING OVERALL DATASET

In [5]:
df= df_import.copy()

In [6]:
df.head(3)

Unnamed: 0  work_year experience_level employment_type  \
0           0       2020               MI              FT   
1           1       2020               SE              FT   
2           2       2020               SE              FT   

                    job_title  salary salary_currency  salary_in_usd  \
0              Data Scientist   70000             EUR          79833   
1  Machine Learning Scientist  260000             USD         260000   
2           Big Data Engineer   85000             GBP         109024   

  employee_residence  remote_ratio company_location company_size  
0                 DE             0               DE            L  
1                 JP             0               JP            S  
2                 GB            50               GB            M

In [7]:
df.shape

(607, 12)

There are 607 samples collected. The dataset has 607 rows and 12 columns

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 607 entries, 0 to 606
Data columns (total 12 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   Unnamed: 0          607 non-null    int64 
 1   work_year           607 non-null    int64 
 2   experience_level    607 non-null    object
 3   employment_type     607 non-null    object
 4   job_title           607 non-null    object
 5   salary              607 non-null    int64 
 6   salary_currency     607 non-null    object
 7   salary_in_usd       607 non-null    int64 
 8   employee_residence  607 non-null    object
 9   remote_ratio        607 non-null    int64 
 10  company_location    607 non-null    object
 11  company_size        607 non-null    object
dtypes: int64(5), object(7)
memory usage: 57.0+ KB


From the result above, we can see that there are 607 samples collected through the survey and all 11 columns have 607 non-null values (it is not sure that there are all non-null value in 11 columns, because the non-null values can be filled by other characters, like: ?, 0,.... So we will check later on). The dataset has the right type for each column

In [9]:
df.columns

Index(['Unnamed: 0', 'work_year', 'experience_level', 'employment_type',
       'job_title', 'salary', 'salary_currency', 'salary_in_usd',
       'employee_residence', 'remote_ratio', 'company_location',
       'company_size'],
      dtype='object')

From the result, I have decided to delete some of them since there are some columns that I think I will not use them

In [10]:
df2= df.drop(['Unnamed: 0','salary','salary_currency'],axis='columns')
df2

work_year experience_level employment_type                   job_title  \
0         2020               MI              FT              Data Scientist   
1         2020               SE              FT  Machine Learning Scientist   
2         2020               SE              FT           Big Data Engineer   
3         2020               MI              FT        Product Data Analyst   
4         2020               SE              FT   Machine Learning Engineer   
..         ...              ...             ...                         ...   
602       2022               SE              FT               Data Engineer   
603       2022               SE              FT               Data Engineer   
604       2022               SE              FT                Data Analyst   
605       2022               SE              FT                Data Analyst   
606       2022               MI              FT                AI Scientist   

     salary_in_usd employee_residence  remote_ratio company_location  \
0            79833                 DE             0               DE   
1           260000                 JP             0               JP   
2           109024                 GB            50               GB   
3            20000                 HN             0               HN   
4           150000                 US            50               US   
..             ...                ...           ...              ...   
602         154000                 US           100               US   
603         126000                 US           100               US   
604         129000                 US             0               US   
605         150000                 US           100               US   
606         200000                 IN           100               US   

    company_size  
0              L  
1              S  
2              M  
3              S  
4              L  
..           ...  
602            M  
603            M  
604            M  
605            M  
606            L  

[607 rows x 9 columns]

So now you see that, we have dropped 3 columns so the number of columns now is 9

In [11]:
df2.nunique()

work_year               3
experience_level        4
employment_type         4
job_title              50
salary_in_usd         369
employee_residence     57
remote_ratio            3
company_location       50
company_size            3
dtype: int64

The result tells me that there is too many job_title, emplyee_residence and company_location and I believe that there are some data which is in this 3 columns appearing few times. So I will check whether I am right or not and I will handle this 

In [12]:
df2_jobtitle= df2.groupby('job_title')['job_title'].agg('count').sort_values(ascending=True).to_frame()
df2_jobtitle.columns=['value_count']
df2_jobtitle

value_count
job_title                                            
3D Computer Vision Researcher                       1
NLP Engineer                                        1
Marketing Data Analyst                              1
Machine Learning Manager                            1
Lead Machine Learning Engineer                      1
Head of Machine Learning                            1
Finance Data Analyst                                1
Data Specialist                                     1
Data Analytics Lead                                 1
Staff Data Scientist                                1
Big Data Architect                                  1
Director of Data Engineering                        2
ETL Developer                                       2
Cloud Data Engineer                                 2
Principal Data Analyst                              2
Financial Data Analyst                              2
Product Data Analyst                                2
Machine Learning Infrastructure Engineer            3
Data Science Engineer                               3
Lead Data Scientist                                 3
Lead Data Analyst                                   3
Computer Vision Software Engineer                   3
Principal Data Engineer                             3
Machine Learning Developer                          3
Analytics Engineer                                  4
Data Analytics Engineer                             4
Applied Machine Learning Scientist                  4
Head of Data Science                                4
Applied Data Scientist                              5
Head of Data                                        5
Data Engineering Manager                            5
Business Data Analyst                               5
ML Engineer                                         6
Computer Vision Engineer                            6
BI Data Analyst                                     6
Lead Data Engineer                                  6
Data Science Consultant                             7
Director of Data Science                            7
AI Scientist                                        7
Data Analytics Manager                              7
Principal Data Scientist                            7
Big Data Engineer                                   8
Machine Learning Scientist                          8
Data Architect                                     11
Data Science Manager                               12
Research Scientist                                 16
Machine Learning Engineer                          41
Data Analyst                                       97
Data Engineer                                     132
Data Scientist                                    143

The result shows that there are lots of work_title have just appeared few times. If we just leave it without any handling, it is not good for our work later (building models,...). So the idea is that I will reduce the number of unique values by changing their job_title name if they have appeared less than 8 times

In [13]:
df2_count_job_title= df2['job_title'].value_counts()

In [14]:
df2_job_confirm= df2_count_job_title[df2_count_job_title<8]
df2['job_title']=df2['job_title'].apply(lambda x:'other' if x in df2_job_confirm else x )
df2['job_title'].unique()

array(['Data Scientist', 'Machine Learning Scientist',
       'Big Data Engineer', 'other', 'Machine Learning Engineer',
       'Data Analyst', 'Data Engineer', 'Research Scientist',
       'Data Science Manager', 'Data Architect'], dtype=object)

You can see that from 50 different values in job_title now we only have 10. GOOD GOOD. we will do the same thing with company_location

In [15]:
df2['company_location']
df2_companylocation_count= df2['company_location'].value_counts().sort_values(ascending=True)
df2_companylocation_count

IE      1
VN      1
EE      1
AS      1
DZ      1
MY      1
MD      1
KE      1
CO      1
IR      1
SG      1
MT      1
IL      1
UA      1
IQ      1
RO      1
HR      1
NZ      1
HU      1
CL      1
HN      1
CH      2
BE      2
CN      2
NG      2
RU      2
CZ      2
IT      2
SI      2
DK      3
TR      3
AU      3
MX      3
AE      3
BR      3
PK      3
LU      3
NL      4
PT      4
PL      4
AT      4
JP      6
GR     11
ES     14
FR     15
IN     24
DE     28
CA     30
GB     47
US    355
Name: company_location, dtype: int64

You can see that there are lots of countries which appeared few times, so the idea is that I will reduce the number of unique values by changing their company_location if they have appeared less than 5 times

In [16]:
df2_companylocation_verified=df2_companylocation_count[df2_companylocation_count <5]
df2['company_location']=df2['company_location'].apply(lambda x: 'other' if x in  df2_companylocation_verified else x)
df2['company_location'].nunique()

10

You can see that we now only have 10 unique values instead of 50 different values

In [17]:
df2['employee_residence'].value_counts().sort_values(ascending=True)

CH      1
MY      1
TN      1
IE      1
DZ      1
AR      1
CZ      1
JE      1
LU      1
PR      1
RS      1
EE      1
CL      1
HK      1
KE      1
MD      1
CO      1
IR      1
CN      1
MT      1
UA      1
IQ      1
HN      1
BG      1
HR      1
PH      1
BO      1
NZ      1
MX      2
HU      2
DK      2
SI      2
SG      2
BE      2
RO      2
NG      2
AU      3
TR      3
VN      3
AT      3
AE      3
PL      4
RU      4
IT      4
NL      5
PK      6
BR      6
PT      6
JP      7
GR     13
ES     15
FR     18
DE     25
CA     29
IN     30
GB     44
US    332
Name: employee_residence, dtype: int64

There are lots of employee_residence which appeared few times. So we should reduce the number of unique values in order to make the process of building model easier later on. I will fillter all employee_residence which have value_counts less than 6 into 'others'.

In [18]:
df2_employee_residence_count=df2['employee_residence'].value_counts() 
df2_employee_residence_count_verified=df2_employee_residence_count[df2_employee_residence_count<6]
df2_employee_residence_count_verified



NL    5
PL    4
IT    4
RU    4
AE    3
AT    3
VN    3
TR    3
AU    3
RO    2
BE    2
SG    2
SI    2
DK    2
HU    2
NG    2
MX    2
BO    1
MY    1
TN    1
IE    1
DZ    1
AR    1
CZ    1
JE    1
LU    1
PR    1
RS    1
EE    1
CL    1
HK    1
KE    1
MD    1
CO    1
IR    1
CN    1
MT    1
UA    1
IQ    1
HN    1
BG    1
HR    1
PH    1
NZ    1
CH    1
Name: employee_residence, dtype: int64

In [19]:
df2['employee_residence']=df2['employee_residence'].apply(lambda x: 'others' if x in df2_employee_residence_count_verified else x)
df2['employee_residence'].replace('others','other',inplace=True)


In [20]:
df2['employee_residence'].nunique()

13

There are only 13 unique values in employee_residence now

# Cleaning and Handling data

In [21]:
df3= df2.copy()

In [22]:
df3.isna().sum()

work_year             0
experience_level      0
employment_type       0
job_title             0
salary_in_usd         0
employee_residence    0
remote_ratio          0
company_location      0
company_size          0
dtype: int64

There is no None value in this dataset, but i will check more carefully in case none value is presented in different charaters.

In [23]:
def unique(df):
    unique_values=[]
    for i in df.columns:
        unique_values.append(i)
        unique_values.append(df[i].unique())
    return unique_values


In [24]:
unique(df3)

['work_year',
 array([2020, 2021, 2022], dtype=int64),
 'experience_level',
 array(['MI', 'SE', 'EN', 'EX'], dtype=object),
 'employment_type',
 array(['FT', 'CT', 'PT', 'FL'], dtype=object),
 'job_title',
 array(['Data Scientist', 'Machine Learning Scientist',
        'Big Data Engineer', 'other', 'Machine Learning Engineer',
        'Data Analyst', 'Data Engineer', 'Research Scientist',
        'Data Science Manager', 'Data Architect'], dtype=object),
 'salary_in_usd',
 array([ 79833, 260000, 109024,  20000, 150000,  72000, 190000,  35735,
        135000, 125000,  51321,  40481,  39916,  87000,  85000,   8000,
         41689, 114047,   5707,  56000,  43331,   6072,  47899,  98000,
        115000, 325000,  42000,  33511, 100000, 117104,  59303,  70000,
         68428, 450000,  46759,  74130, 103000, 250000,  10000, 138000,
         45760,  50180, 106000, 112872,  15966,  76958, 188000, 105000,
         70139,  91000,  45896,  54742,  60000, 148261,  38776, 118000,
        120000, 1383

From the reruslt, we can jump to the conclusion that there is no None value

There are some words presented in acronym, so I write them in the full words  so as to be easy to understand when we visualize them into graphs or charts

In [25]:
df3['experience_level'].unique()

array(['MI', 'SE', 'EN', 'EX'], dtype=object)

In [26]:
df3['experience_level']= df3['experience_level'].map({'MI':'Mid Level','SE':'Senior Level','EN':'Entry Level','EX':'Executive Level'},na_action='ignore')

In [27]:
df3['experience_level'].unique()

array(['Mid Level', 'Senior Level', 'Entry Level', 'Executive Level'],
      dtype=object)

We have completed converting experience_level written in acronym to full-word. Now we do the same thing to employment_type, remote_ratio, company_size

In [28]:
df3['employment_type'].unique()

array(['FT', 'CT', 'PT', 'FL'], dtype=object)

In [29]:
df3['employment_type'].replace({'FT':'Full time','PT':'Part time','CT':'Contract','FL':'Freelance'},inplace=True)

In [30]:
df3['employment_type'].unique()

array(['Full time', 'Contract', 'Part time', 'Freelance'], dtype=object)

In [31]:
df3['remote_ratio'].unique()

array([  0,  50, 100], dtype=int64)

In [32]:
df3['remote_ratio']=df3['remote_ratio'].map({0:'On-site',50:'Hybrid',100:'Remote'})
df3['remote_ratio'].unique()

array(['On-site', 'Hybrid', 'Remote'], dtype=object)

In [33]:
df3['company_size'].unique()


array(['L', 'S', 'M'], dtype=object)

In [34]:
df3['company_size'].replace({'L':'Large','S':'Small','M':'Medium'},inplace=True)
df3['company_size'].unique()

array(['Large', 'Small', 'Medium'], dtype=object)

Now, we have done. In the next section we will deep into analysis. 

# EDA - Univariate analysis

In [35]:
df4 =df3.copy()
df4

work_year experience_level employment_type                   job_title  \
0         2020        Mid Level       Full time              Data Scientist   
1         2020     Senior Level       Full time  Machine Learning Scientist   
2         2020     Senior Level       Full time           Big Data Engineer   
3         2020        Mid Level       Full time                       other   
4         2020     Senior Level       Full time   Machine Learning Engineer   
..         ...              ...             ...                         ...   
602       2022     Senior Level       Full time               Data Engineer   
603       2022     Senior Level       Full time               Data Engineer   
604       2022     Senior Level       Full time                Data Analyst   
605       2022     Senior Level       Full time                Data Analyst   
606       2022        Mid Level       Full time                       other   

     salary_in_usd employee_residence remote_ratio company_location  \
0            79833                 DE      On-site               DE   
1           260000                 JP      On-site               JP   
2           109024                 GB       Hybrid               GB   
3            20000              other      On-site            other   
4           150000                 US       Hybrid               US   
..             ...                ...          ...              ...   
602         154000                 US       Remote               US   
603         126000                 US       Remote               US   
604         129000                 US      On-site               US   
605         150000                 US       Remote               US   
606         200000                 IN       Remote               US   

    company_size  
0          Large  
1          Small  
2         Medium  
3          Small  
4          Large  
..           ...  
602       Medium  
603       Medium  
604       Medium  
605       Medium  
606        Large  

[607 rows x 9 columns]

In [36]:
df4_dropprice= df4.drop('salary_in_usd',axis=1)

In [37]:
def uniravite(df):
      for i in df.columns:
        x = df[i].value_counts().reset_index()
        x.columns=['a','b']
        x['a']=x['a'].astype('str')
        fig = make_subplots(rows=2, cols=1,specs=[[{"type": "bar"}],[{"type": "pie"}]],
)

        fig.add_trace(go.Bar(x=x['a'],y=x['b'],text=x['b']),
              row=1, col=1)

        fig.add_trace(go.Pie(labels=x['a'],values=x['b']),
              row=2, col=1)


        fig.update_layout(height=700, showlegend=True)
        q=fig.show()

        q
      return q
        

In [38]:
from plotly.subplots import make_subplots
import plotly.graph_objects as go

In [39]:
uniravite(df4_dropprice)

from the results, we have conclusions for the distribution of data in each columns:
1. Work_year: The majority of samples collected is from 2022 and 2021, which are 318 samples and 217 samples respectively. in 2020, there are only  72 samples. Therefore, we can say that the salary that we analyze is almost from 2022 and 2021.
2. Exerience_level: From the results you can see that the number of people who is executive level is really small. However, this in Senior level is significant high. The number of people who is mid level is a bit smaller than this in senior level.
3. Employment_type: most of employee have worked full-time and the number of people working part-time, contract, or freelance is very small.
4. Job_title: people mostly work as Data Sicentist, Data Engineer, and Data Analyst. The number of peole working in other fields is greatly small. So we can say that the majority of samples collected focus on 3 types of works: Data Sicentist, Data Engineer, Data Analyst.
5. Employee_residence: Staff who is interviewed is mostly from US, which account for 54.7%.
6. Remote_ratio: People tend to work remotely, which make up 62.8%. Staff work on-site, and hybrid are just around 20.9%, and 16.3%.
7. Company_location: Most companies in the samples are from US.
8. Company_size: the percentage of large companies is about 32.6% and this of medium enterprises is around 54%, which is very high.



In [40]:
fig= px.histogram(df4,x=df4['salary_in_usd'])
fig.show()

Staff  working in these types of fields mostly has the salary from 60k to 160k. There are few people have high salary from 400k to 600k per year.

When I do some research about the salary of Data Scientist, I have realized that there are people who have the salary up to 580k per year. As the result, I do not think this data has outliners.

# EDA: Multivariate analysis

In [41]:
df4

work_year experience_level employment_type                   job_title  \
0         2020        Mid Level       Full time              Data Scientist   
1         2020     Senior Level       Full time  Machine Learning Scientist   
2         2020     Senior Level       Full time           Big Data Engineer   
3         2020        Mid Level       Full time                       other   
4         2020     Senior Level       Full time   Machine Learning Engineer   
..         ...              ...             ...                         ...   
602       2022     Senior Level       Full time               Data Engineer   
603       2022     Senior Level       Full time               Data Engineer   
604       2022     Senior Level       Full time                Data Analyst   
605       2022     Senior Level       Full time                Data Analyst   
606       2022        Mid Level       Full time                       other   

     salary_in_usd employee_residence remote_ratio company_location  \
0            79833                 DE      On-site               DE   
1           260000                 JP      On-site               JP   
2           109024                 GB       Hybrid               GB   
3            20000              other      On-site            other   
4           150000                 US       Hybrid               US   
..             ...                ...          ...              ...   
602         154000                 US       Remote               US   
603         126000                 US       Remote               US   
604         129000                 US      On-site               US   
605         150000                 US       Remote               US   
606         200000                 IN       Remote               US   

    company_size  
0          Large  
1          Small  
2         Medium  
3          Small  
4          Large  
..           ...  
602       Medium  
603       Medium  
604       Medium  
605       Medium  
606        Large  

[607 rows x 9 columns]

Since there are several columns in this dataset, so it is very difficult and time-consuming to analyze all of these columns in multivariate analysis. Therefore, I will just concentrate on analyzing some relevant factors that I think I will use to build models.

In [42]:
df4['work_year']=df4['work_year'].astype('str')
df4_averagesalarytoyear= df4.groupby('work_year')['salary_in_usd'].agg('mean')
df4_averagesalarytoyear=df4_averagesalarytoyear.reset_index()
df4_averagesalarytoyear

work_year  salary_in_usd
0      2020   95813.000000
1      2021   99853.792627
2      2022  124522.006289

In [43]:
fig=px.bar(df4_averagesalarytoyear,x='work_year',y='salary_in_usd',text=df4_averagesalarytoyear['salary_in_usd'],title='Work Year Salary')
fig.update_layout(xaxis_title='Work Year',yaxis_title='Average Salary In USD')
fig.show()

Overall, we can see that the salary that the employee earned in 2022 is higher than 2020 and 2021

In [44]:
df4.columns

Index(['work_year', 'experience_level', 'employment_type', 'job_title',
       'salary_in_usd', 'employee_residence', 'remote_ratio',
       'company_location', 'company_size'],
      dtype='object')

In [45]:
df4_jobtitleoversalary=df4.groupby('job_title')['salary_in_usd'].agg('mean')
df4_jobtitleoversalary=df4_jobtitleoversalary.reset_index()


In [46]:
fig=px.bar(x=df4_jobtitleoversalary['job_title'],y=df4_jobtitleoversalary['salary_in_usd'],text=df4_jobtitleoversalary['salary_in_usd'],title='Job title Salary')
fig.update_layout( xaxis_title="Job",yaxis_title="Average Salary",legend_title="Legend Title")
fig.show()

Suprisingly, Data Architect is the job that have the highest average wage,compared to the others

In [47]:
df4_jobtitle_workyear_salary=df4.groupby(['job_title','work_year'])['salary_in_usd'].mean().reset_index()
df4_jobtitle_workyear_salary

job_title work_year  salary_in_usd
0            Big Data Engineer      2020   97690.333333
1            Big Data Engineer      2021   24544.200000
2                 Data Analyst      2020   45547.285714
3                 Data Analyst      2021   79505.411765
4                 Data Analyst      2022  100550.739726
5               Data Architect      2021  166666.666667
6               Data Architect      2022  182076.625000
7                Data Engineer      2020   88162.000000
8                Data Engineer      2021   83202.531250
9                Data Engineer      2022  126375.696629
10        Data Science Manager      2020  190200.000000
11        Data Science Manager      2021  143126.500000
12        Data Science Manager      2022  170196.600000
13              Data Scientist      2020   85970.523810
14              Data Scientist      2021   70671.733333
15              Data Scientist      2022  136172.090909
16   Machine Learning Engineer      2020  125389.800000
17   Machine Learning Engineer      2021   74611.222222
18   Machine Learning Engineer      2022  129451.944444
19  Machine Learning Scientist      2020  260000.000000
20  Machine Learning Scientist      2021  145500.000000
21  Machine Learning Scientist      2022  141766.666667
22          Research Scientist      2020  246000.000000
23          Research Scientist      2021   83003.600000
24          Research Scientist      2022  105569.000000
25                       other      2020   92491.523810
26                       other      2021  132953.948052
27                       other      2022  122921.926829

In [48]:
fig=px.bar(df4_jobtitle_workyear_salary,x='job_title',y='salary_in_usd',color='work_year',text='salary_in_usd',barmode='group',title='Average salary_job department with work year')
fig.update_layout(yaxis_title='Average Salary in USD',xaxis_title='Job Title')
fig.show()

From the results, we can see that in 2022 the average salary in usd is  higher than this in 2021 and 2000 in almost job title, except Machine learning Scientist, Research Scietist, Data Science Manager... Suprisingly, in 2020 Machine Learning Scientist and Research Scientist have the average Salary in USD, which make me a little confused, and Data Scientist seems to have the small Average Salary In USD although it is a factor that appears so many times in the samples. So I will search more deeply into them.

In [49]:
df4_searchdeeply =df4.groupby('job_title').size().sort_values(ascending=True).reset_index()
df4_searchdeeply.columns=['job_title','times appeaaring']
df4_searchdeeply[df4_searchdeeply['job_title'].isin(['Machine Learning Scientist','Data Science Manager','Data Scientist'])]


job_title  times appeaaring
1  Machine Learning Scientist                 8
3        Data Science Manager                12
9              Data Scientist               143

In [50]:
df4_datascience= df4[df4['job_title']=='Data Scientist'][['job_title','salary_in_usd',]]
df4_datascience=df4_datascience.reset_index()
df4_machinescientest= df4[df4['job_title']=='Machine Learning Scientist'][['job_title','salary_in_usd']].reset_index()
df4_Datamanager= df4[df4['job_title']=='Data Science Manager'][['job_title','salary_in_usd']].reset_index()

We have already done the preperation for analyzing more deeply into 3 types of works. Now we will plot so we can have a visualization.

In [51]:
fig = make_subplots(
    rows=3, cols=1,subplot_titles=('Data Science','Machine Learning Scientist','Data Science Manager')
)

fig.add_trace(go.Histogram(x=df4_datascience['salary_in_usd'],),
              row=1, col=1)

fig.add_trace(go.Histogram(x=df4_machinescientest['salary_in_usd']),
              row=2, col=1)

fig.add_trace(go.Histogram(x=df4_Datamanager['salary_in_usd']),
              row=3, col=1)


fig.update_layout(height=700, showlegend=False)

We can conclude that although data scientist job has appeared so many times but the employee has earned around 50k to 150k, which is much smaller than the salary can be earned by staff working as Machine Learning Scientist, and Data Science Manager.

In [52]:
df4

work_year experience_level employment_type                   job_title  \
0        2020        Mid Level       Full time              Data Scientist   
1        2020     Senior Level       Full time  Machine Learning Scientist   
2        2020     Senior Level       Full time           Big Data Engineer   
3        2020        Mid Level       Full time                       other   
4        2020     Senior Level       Full time   Machine Learning Engineer   
..        ...              ...             ...                         ...   
602      2022     Senior Level       Full time               Data Engineer   
603      2022     Senior Level       Full time               Data Engineer   
604      2022     Senior Level       Full time                Data Analyst   
605      2022     Senior Level       Full time                Data Analyst   
606      2022        Mid Level       Full time                       other   

     salary_in_usd employee_residence remote_ratio company_location  \
0            79833                 DE      On-site               DE   
1           260000                 JP      On-site               JP   
2           109024                 GB       Hybrid               GB   
3            20000              other      On-site            other   
4           150000                 US       Hybrid               US   
..             ...                ...          ...              ...   
602         154000                 US       Remote               US   
603         126000                 US       Remote               US   
604         129000                 US      On-site               US   
605         150000                 US       Remote               US   
606         200000                 IN       Remote               US   

    company_size  
0          Large  
1          Small  
2         Medium  
3          Small  
4          Large  
..           ...  
602       Medium  
603       Medium  
604       Medium  
605       Medium  
606        Large  

[607 rows x 9 columns]

Now we move to analyze other columns.

In [53]:
df4_yearvsexperience= df4.groupby(['work_year','experience_level'])['salary_in_usd'].mean().reset_index()
df4_yearvsexperience['work_year']=df4_yearvsexperience['work_year'].astype('str')
df4_yearvsexperience

work_year experience_level  salary_in_usd
0       2020      Entry Level   63648.600000
1       2020  Executive Level  202416.500000
2       2020        Mid Level   85950.062500
3       2020     Senior Level  137240.500000
4       2021      Entry Level   59101.021277
5       2021  Executive Level  223752.727273
6       2021        Mid Level   85490.088889
7       2021     Senior Level  126596.188406
8       2022      Entry Level   65423.428571
9       2022  Executive Level  178313.846154
10      2022        Mid Level   91193.956044
11      2022     Senior Level  143043.398964

In [54]:
fig= px.bar(df4_yearvsexperience, x='work_year',y='salary_in_usd',color='experience_level',barmode='group')
fig.show()

It is easy to understand that if staff holds a higher position in companies, they can have higher salary. 

In [55]:
df4_jobvsexperience= df4.groupby(['job_title','experience_level'])['salary_in_usd'].mean().reset_index()
df4_jobvsexperience['work_year']=df4_yearvsexperience['work_year'].astype('str')
df4_jobvsexperience

job_title experience_level  salary_in_usd work_year
0            Big Data Engineer      Entry Level   30703.333333      2020
1            Big Data Engineer        Mid Level   33537.000000      2020
2            Big Data Engineer     Senior Level  111535.500000      2020
3                 Data Analyst      Entry Level   53960.666667      2020
4                 Data Analyst  Executive Level  120000.000000      2021
5                 Data Analyst        Mid Level   71699.206897      2021
6                 Data Analyst     Senior Level  111922.629630      2021
7               Data Architect        Mid Level  166666.666667      2021
8               Data Architect     Senior Level  182076.625000      2022
9                Data Engineer      Entry Level   58933.500000      2022
10               Data Engineer  Executive Level  245500.000000      2022
11               Data Engineer        Mid Level   85985.660377      2022
12               Data Engineer     Senior Level  137035.841270       NaN
13        Data Science Manager        Mid Level  200000.000000       NaN
14        Data Science Manager     Senior Level  149994.200000       NaN
15              Data Scientist      Entry Level   55330.909091       NaN
16              Data Scientist        Mid Level   82039.133333       NaN
17              Data Scientist     Senior Level  152971.016393       NaN
18   Machine Learning Engineer      Entry Level   86996.333333       NaN
19   Machine Learning Engineer        Mid Level   74466.583333       NaN
20   Machine Learning Engineer     Senior Level  131176.000000       NaN
21  Machine Learning Scientist      Entry Level  225000.000000       NaN
22  Machine Learning Scientist        Mid Level  109325.000000       NaN
23  Machine Learning Scientist     Senior Level  201666.666667       NaN
24          Research Scientist      Entry Level   86132.000000       NaN
25          Research Scientist        Mid Level  136498.142857       NaN
26          Research Scientist     Senior Level   88859.400000       NaN
27                       other      Entry Level   56319.880000       NaN
28                       other  Executive Level  198109.650000       NaN
29                       other        Mid Level   97432.450000       NaN
30                       other     Senior Level  147260.870370       NaN

In [56]:
fig= px.bar(df4_jobvsexperience, x='job_title',y='salary_in_usd',color='experience_level',barmode='group')
fig.show()

From the chart, we can see that it is unsure that  all jobs will have executive position. Excecutive level is just available in some jobs such as: Data Analyst, Data Engineer, and some other jobs. In my opinion, I do not 100% believe that this hypothesis is true since the data can be skewed. However, you can see from the chart that the higher positon staff have, the higher salary that they can earn from enterprises, but in Data Science Manager the data show that Mid Level employee get more earnings when comparing to Senior Level staff, which is ridiculous. So I will do more research in this field

In [57]:
df4[df4['job_title']=='Data Science Manager'][['experience_level','salary_in_usd']]

experience_level  salary_in_usd
67      Senior Level         190200
141     Senior Level         240000
154     Senior Level         144000
190     Senior Level         152000
193     Senior Level         174000
198     Senior Level          54094
285     Senior Level          94665
350     Senior Level         161342
351     Senior Level         137141
421        Mid Level         241000
422        Mid Level         159000
516     Senior Level         152500

After asking some people who work in this type of job, I see that staff working as Data science Manager and holding mid position and working in US can not have a salary that is higher than 200k, so I will filter some mid level employees whose salary is higher than 200k. Moreover, through research, I have found that senior employees working as Data Science Manager can not earn the salary which is under 160k.

In [58]:
df4= df4[~((df4['experience_level']=='Mid Level') & (df4['salary_in_usd']>200000) & (df4['job_title']=='Data Science Manager'))]


When analyzing data, I sill wonder that there are some senior staff have the very low salary, maybe the salary they can get is much smaller than mid level staff. However, I have realized that this salary also depends on company_location which means if the staff working in India will have lower earnings than their counterpart working in US.

In [59]:
df4_jobvsexperience1= df4.groupby(['job_title','experience_level'])['salary_in_usd'].mean().reset_index()
df4_jobvsexperience1['work_year']=df4_yearvsexperience['work_year'].astype('str')
fig= px.bar(df4_jobvsexperience1, x='job_title',y='salary_in_usd',color='experience_level',barmode='group')
fig.show()

After fitering data, I have still saw that some staff who holds higher position gets lower wages, such as: Machine Learning Scientist, Data Science Manager. However, as I mention before, the salary can depends on so many factors, like: company_position, remote_ratio... So for now, It is acceptable. 

In [60]:
df4

work_year experience_level employment_type                   job_title  \
0        2020        Mid Level       Full time              Data Scientist   
1        2020     Senior Level       Full time  Machine Learning Scientist   
2        2020     Senior Level       Full time           Big Data Engineer   
3        2020        Mid Level       Full time                       other   
4        2020     Senior Level       Full time   Machine Learning Engineer   
..        ...              ...             ...                         ...   
602      2022     Senior Level       Full time               Data Engineer   
603      2022     Senior Level       Full time               Data Engineer   
604      2022     Senior Level       Full time                Data Analyst   
605      2022     Senior Level       Full time                Data Analyst   
606      2022        Mid Level       Full time                       other   

     salary_in_usd employee_residence remote_ratio company_location  \
0            79833                 DE      On-site               DE   
1           260000                 JP      On-site               JP   
2           109024                 GB       Hybrid               GB   
3            20000              other      On-site            other   
4           150000                 US       Hybrid               US   
..             ...                ...          ...              ...   
602         154000                 US       Remote               US   
603         126000                 US       Remote               US   
604         129000                 US      On-site               US   
605         150000                 US       Remote               US   
606         200000                 IN       Remote               US   

    company_size  
0          Large  
1          Small  
2         Medium  
3          Small  
4          Large  
..           ...  
602       Medium  
603       Medium  
604       Medium  
605       Medium  
606        Large  

[606 rows x 9 columns]

In [61]:
df4_employmenttypevssalary= df4.groupby('employment_type')['salary_in_usd'].agg(['mean','size']).reset_index()
df4_employmenttypevssalary

employment_type           mean  size
0        Contract  184575.000000     5
1       Freelance   48000.000000     4
2       Full time  113250.812606   587
3       Part time   33070.500000    10

In [62]:
fig=px.bar(df4_employmenttypevssalary,x='employment_type',y='mean',hover_data=['size'])
fig.update_layout(xaxis_title= 'Employment type',yaxis_title= 'Average_in_usd')
fig.show()

Suprisingly, Contact employees have higher salary when comparting to full-time employees. I always think that full-time staff will get the highest earnings. However, when looking at data again, I have seen that contact employees have appeared just 5 times in the samples but full-time employees have appeared 587 times. Therefore, I will search more about the distribution of salary in full-time employees.

In [63]:
df4_fulltime= df4[df4['employment_type']=='Full time'][['employment_type','salary_in_usd']]
df4_fulltime

employment_type  salary_in_usd
0         Full time          79833
1         Full time         260000
2         Full time         109024
3         Full time          20000
4         Full time         150000
..              ...            ...
602       Full time         154000
603       Full time         126000
604       Full time         129000
605       Full time         150000
606       Full time         200000

[587 rows x 2 columns]

In [64]:
fig=px.histogram(df4_fulltime,x='salary_in_usd',title='Distribution of full-time salary')
fig.show()

So the result can be explained what make me confused. Although full-time employees have apeared several times in the dataset, but their salary is not high, which focus from 40k to 140k. 

When doing this research, I think that we should find and detect some outliners since I see so many outliners in salary_in_usd.

In [65]:
df4

work_year experience_level employment_type                   job_title  \
0        2020        Mid Level       Full time              Data Scientist   
1        2020     Senior Level       Full time  Machine Learning Scientist   
2        2020     Senior Level       Full time           Big Data Engineer   
3        2020        Mid Level       Full time                       other   
4        2020     Senior Level       Full time   Machine Learning Engineer   
..        ...              ...             ...                         ...   
602      2022     Senior Level       Full time               Data Engineer   
603      2022     Senior Level       Full time               Data Engineer   
604      2022     Senior Level       Full time                Data Analyst   
605      2022     Senior Level       Full time                Data Analyst   
606      2022        Mid Level       Full time                       other   

     salary_in_usd employee_residence remote_ratio company_location  \
0            79833                 DE      On-site               DE   
1           260000                 JP      On-site               JP   
2           109024                 GB       Hybrid               GB   
3            20000              other      On-site            other   
4           150000                 US       Hybrid               US   
..             ...                ...          ...              ...   
602         154000                 US       Remote               US   
603         126000                 US       Remote               US   
604         129000                 US      On-site               US   
605         150000                 US       Remote               US   
606         200000                 IN       Remote               US   

    company_size  
0          Large  
1          Small  
2         Medium  
3          Small  
4          Large  
..           ...  
602       Medium  
603       Medium  
604       Medium  
605       Medium  
606        Large  

[606 rows x 9 columns]

In [66]:
fig=px.histogram(df4,x='salary_in_usd')
fig.show()

I will deeply analyze employees whose salary is higher than 240k

In [67]:
df4[df4['salary_in_usd']>240000]

work_year experience_level employment_type                   job_title  \
1        2020     Senior Level       Full time  Machine Learning Scientist   
25       2020  Executive Level       Full time                       other   
33       2020        Mid Level       Full time          Research Scientist   
37       2020      Entry Level       Full time   Machine Learning Engineer   
63       2020     Senior Level       Full time              Data Scientist   
78       2021        Mid Level        Contract                       other   
93       2021     Senior Level       Full time                       other   
97       2021        Mid Level       Full time                       other   
157      2021        Mid Level       Full time                       other   
167      2021  Executive Level       Full time                       other   
225      2021  Executive Level        Contract                       other   
231      2021     Senior Level       Full time                       other   
252      2021  Executive Level       Full time                       other   
309      2022  Executive Level       Full time               Data Engineer   
337      2022     Senior Level       Full time               Data Engineer   
416      2022     Senior Level       Full time              Data Scientist   
482      2022  Executive Level       Full time               Data Engineer   
519      2022     Senior Level       Full time                       other   
523      2022     Senior Level       Full time                       other   
534      2022     Senior Level       Full time              Data Architect   

     salary_in_usd employee_residence remote_ratio company_location  \
1           260000                 JP      On-site               JP   
25          325000                 US       Remote               US   
33          450000                 US      On-site               US   
37          250000                 US       Hybrid               US   
63          412000                 US       Remote               US   
78          270000                 US       Remote               US   
93          276000                 US      On-site               US   
97          450000                 US       Remote               US   
157         423000                 US       Hybrid               US   
167         250000                 US      On-site               US   
225         416000                 US       Remote               US   
231         256000                 US       Remote               US   
252         600000                 US       Remote               US   
309         242000                 US       Remote               US   
337         243900                 US       Remote               US   
416         260000                 US       Remote               US   
482         324000                 US       Remote               US   
519         380000                 US       Remote               US   
523         405000                 US       Remote               US   
534         266400                 US       Remote               US   

    company_size  
1          Small  
25         Large  
33        Medium  
37         Large  
63         Large  
78         Large  
93         Large  
97         Large  
157        Large  
167        Large  
225        Small  
231        Small  
252        Large  
309       Medium  
337       Medium  
416       Medium  
482       Medium  
519        Large  
523        Large  
534       Medium

From the results, I suppose that there is only one data that is not logical. This is the data whose index is 63. This staff work as Data Scientist, he/she is senior level and work full-time and his/her salary is above 400k, which is very strange. Thus, I decide to eliminate it.

In [68]:
df4= df4[~((df4['job_title']=='Data Scientist') & (df4['salary_in_usd']== 412000))]
df4

work_year experience_level employment_type                   job_title  \
0        2020        Mid Level       Full time              Data Scientist   
1        2020     Senior Level       Full time  Machine Learning Scientist   
2        2020     Senior Level       Full time           Big Data Engineer   
3        2020        Mid Level       Full time                       other   
4        2020     Senior Level       Full time   Machine Learning Engineer   
..        ...              ...             ...                         ...   
602      2022     Senior Level       Full time               Data Engineer   
603      2022     Senior Level       Full time               Data Engineer   
604      2022     Senior Level       Full time                Data Analyst   
605      2022     Senior Level       Full time                Data Analyst   
606      2022        Mid Level       Full time                       other   

     salary_in_usd employee_residence remote_ratio company_location  \
0            79833                 DE      On-site               DE   
1           260000                 JP      On-site               JP   
2           109024                 GB       Hybrid               GB   
3            20000              other      On-site            other   
4           150000                 US       Hybrid               US   
..             ...                ...          ...              ...   
602         154000                 US       Remote               US   
603         126000                 US       Remote               US   
604         129000                 US      On-site               US   
605         150000                 US       Remote               US   
606         200000                 IN       Remote               US   

    company_size  
0          Large  
1          Small  
2         Medium  
3          Small  
4          Large  
..           ...  
602       Medium  
603       Medium  
604       Medium  
605       Medium  
606        Large  

[605 rows x 9 columns]

In [69]:
df4_jobandemployee=df4.groupby(['job_title','employee_residence'])['salary_in_usd'].mean().reset_index()
df4_jobandemployee

job_title employee_residence  salary_in_usd
0   Big Data Engineer                 ES   60000.000000
1   Big Data Engineer                 GB  109024.000000
2   Big Data Engineer                 IN   14907.000000
3   Big Data Engineer                 US   70000.000000
4   Big Data Engineer              other   66023.500000
..                ...                ...            ...
65              other                 JP  103025.000000
66              other                 PK   16000.000000
67              other                 PT   39283.600000
68              other                 US  187807.912281
69              other              other   85757.791667

[70 rows x 3 columns]

In [70]:
fig=px.bar(df4_jobandemployee,x='job_title',y='salary_in_usd',color='employee_residence',barmode='group')
fig.show()

The American people tend to get higher salary than the others in all jobs. In the samples, the British people get the highest salary in Big Data Engineer, while the Japanese people earn more wages in Machine Learning Scientist.

In [71]:
df4_employeevsresidence =df4.groupby(['employee_residence','experience_level']).size().reset_index()
df4_employeevsresidence.columns=['employee_residence','experience_level','number_count']
df4_employeevsresidence


employee_residence experience_level  number_count
0                  BR      Entry Level             1
1                  BR        Mid Level             1
2                  BR     Senior Level             4
3                  CA      Entry Level             3
4                  CA  Executive Level             2
5                  CA        Mid Level            10
6                  CA     Senior Level            14
7                  DE      Entry Level             8
8                  DE  Executive Level             2
9                  DE        Mid Level             8
10                 DE     Senior Level             7
11                 ES      Entry Level             1
12                 ES  Executive Level             1
13                 ES        Mid Level            12
14                 ES     Senior Level             1
15                 FR      Entry Level             5
16                 FR  Executive Level             1
17                 FR        Mid Level             8
18                 FR     Senior Level             4
19                 GB      Entry Level             5
20                 GB        Mid Level            30
21                 GB     Senior Level             9
22                 GR        Mid Level            10
23                 GR     Senior Level             3
24                 IN      Entry Level            11
25                 IN  Executive Level             2
26                 IN        Mid Level            12
27                 IN     Senior Level             5
28                 JP      Entry Level             2
29                 JP        Mid Level             3
30                 JP     Senior Level             2
31                 PK      Entry Level             4
32                 PK        Mid Level             2
33                 PT      Entry Level             2
34                 PT        Mid Level             3
35                 PT     Senior Level             1
36                 US      Entry Level            29
37                 US  Executive Level            15
38                 US        Mid Level            76
39                 US     Senior Level           210
40              other      Entry Level            17
41              other  Executive Level             3
42              other        Mid Level            37
43              other     Senior Level            19

In [72]:
fig= px.bar(df4_employeevsresidence,x='employee_residence',y='number_count',color='experience_level',barmode='group')
fig.show()

From the results, we can see that there is few executive employees in the samples. Mid Level and Senior Level staff are two factors making up the most figure

In [73]:
df4

work_year experience_level employment_type                   job_title  \
0        2020        Mid Level       Full time              Data Scientist   
1        2020     Senior Level       Full time  Machine Learning Scientist   
2        2020     Senior Level       Full time           Big Data Engineer   
3        2020        Mid Level       Full time                       other   
4        2020     Senior Level       Full time   Machine Learning Engineer   
..        ...              ...             ...                         ...   
602      2022     Senior Level       Full time               Data Engineer   
603      2022     Senior Level       Full time               Data Engineer   
604      2022     Senior Level       Full time                Data Analyst   
605      2022     Senior Level       Full time                Data Analyst   
606      2022        Mid Level       Full time                       other   

     salary_in_usd employee_residence remote_ratio company_location  \
0            79833                 DE      On-site               DE   
1           260000                 JP      On-site               JP   
2           109024                 GB       Hybrid               GB   
3            20000              other      On-site            other   
4           150000                 US       Hybrid               US   
..             ...                ...          ...              ...   
602         154000                 US       Remote               US   
603         126000                 US       Remote               US   
604         129000                 US      On-site               US   
605         150000                 US       Remote               US   
606         200000                 IN       Remote               US   

    company_size  
0          Large  
1          Small  
2         Medium  
3          Small  
4          Large  
..           ...  
602       Medium  
603       Medium  
604       Medium  
605       Medium  
606        Large  

[605 rows x 9 columns]

In [74]:
df4_companylocationandjobtitle =df4.groupby(['company_location','job_title'])['salary_in_usd'].agg(['mean','count']).reset_index()
df4_companylocationandjobtitle


company_location                   job_title           mean  count
0                CA                Data Analyst   81487.333333      9
1                CA              Data Architect  141550.000000      2
2                CA              Data Scientist   77787.000000      5
3                CA  Machine Learning Scientist  225000.000000      1
4                CA          Research Scientist  115788.333333      3
5                CA                       other  101692.600000     10
6                DE                Data Analyst   63831.000000      1
7                DE               Data Engineer   58186.750000      4
8                DE              Data Scientist   69640.142857      7
9                DE   Machine Learning Engineer   73079.750000      4
10               DE                       other  101372.000000     12
11               ES                Data Analyst   29098.000000      3
12               ES               Data Engineer   70070.500000      4
13               ES              Data Scientist   41136.666667      3
14               ES   Machine Learning Engineer   47282.000000      1
15               ES                       other   68191.333333      3
16               FR                Data Analyst   52930.500000      2
17               FR               Data Engineer   69143.000000      2
18               FR        Data Science Manager  152000.000000      1
19               FR              Data Scientist   50085.571429      7
20               FR          Research Scientist   70938.000000      3
21               GB           Big Data Engineer  111535.500000      2
22               GB                Data Analyst   50875.600000      5
23               GB               Data Engineer   80305.750000     20
24               GB              Data Scientist   88177.363636     11
25               GB   Machine Learning Engineer   86597.000000      3
26               GB          Research Scientist   82528.000000      1
27               GB                       other   87713.800000      5
28               GR                Data Analyst   32313.333333      3
29               GR               Data Engineer   65636.200000      5
30               GR                       other   50034.333333      3
31               IN           Big Data Engineer   19419.500000      2
32               IN                Data Analyst    6072.000000      1
33               IN               Data Engineer   26032.500000      2
34               IN        Data Science Manager   94665.000000      1
35               IN              Data Scientist   26108.250000      8
36               IN   Machine Learning Engineer   36869.000000      3
37               IN                       other   24978.285714      7
38               JP               Data Engineer   41689.000000      1
39               JP   Machine Learning Engineer   74000.000000      1
40               JP  Machine Learning Scientist  260000.000000      1
41               JP                       other  103025.000000      3
42               US           Big Data Engineer   70000.000000      1
43               US                Data Analyst  107674.154930     71
44               US              Data Architect  185945.888889      9
45               US               Data Engineer  139724.682353     85
46               US        Data Science Manager  156919.666667      9
47               US              Data Scientist  139876.108434     83
48               US   Machine Learning Engineer  169055.750000     16
49               US  Machine Learning Scientist  154060.000000      5
50               US          Research Scientist  191000.000000      4
51               US                       other  170863.657143     70
52            other           Big Data Engineer   27960.666667      3
53            other                Data Analyst    9000.000000      2
54            other               Data Engineer   35970.777778      9
55            other              Data Scientist   51105.611111     18
56           

In [75]:
fig= px.bar(df4_companylocationandjobtitle,x='company_location',y='mean',color='job_title',barmode='group',hover_data=['count'])
fig.update_layout(yaxis_title= 'Average Salary')
fig.show()

The average salary is almost the same in Canada and US. while we can sen that in US there is no great difference in average salary in different jobs, in Canada, Machine Learning Scientist is the well paid job, and this in Japan is alsao Machine Learning Scientist.

In [76]:
df4size= df4.groupby('company_size')['salary_in_usd'].mean().reset_index()
df4size

company_size  salary_in_usd
0        Large  117756.918782
1       Medium  116523.636923
2        Small   77632.674699

In [77]:
fig=px.bar(df4size,x='company_size',y='salary_in_usd',text='salary_in_usd')
fig.show()

Overall, Large and Medium company nearly pay the same salary for their staff, around 120k. 

In [78]:
df4

work_year experience_level employment_type                   job_title  \
0        2020        Mid Level       Full time              Data Scientist   
1        2020     Senior Level       Full time  Machine Learning Scientist   
2        2020     Senior Level       Full time           Big Data Engineer   
3        2020        Mid Level       Full time                       other   
4        2020     Senior Level       Full time   Machine Learning Engineer   
..        ...              ...             ...                         ...   
602      2022     Senior Level       Full time               Data Engineer   
603      2022     Senior Level       Full time               Data Engineer   
604      2022     Senior Level       Full time                Data Analyst   
605      2022     Senior Level       Full time                Data Analyst   
606      2022        Mid Level       Full time                       other   

     salary_in_usd employee_residence remote_ratio company_location  \
0            79833                 DE      On-site               DE   
1           260000                 JP      On-site               JP   
2           109024                 GB       Hybrid               GB   
3            20000              other      On-site            other   
4           150000                 US       Hybrid               US   
..             ...                ...          ...              ...   
602         154000                 US       Remote               US   
603         126000                 US       Remote               US   
604         129000                 US      On-site               US   
605         150000                 US       Remote               US   
606         200000                 IN       Remote               US   

    company_size  
0          Large  
1          Small  
2         Medium  
3          Small  
4          Large  
..           ...  
602       Medium  
603       Medium  
604       Medium  
605       Medium  
606        Large  

[605 rows x 9 columns]

In [79]:
df4remoteandjob= df4.groupby(['job_title','remote_ratio'])['salary_in_usd'].agg(['mean','size']).reset_index()
df4remoteandjob

job_title remote_ratio           mean  size
0            Big Data Engineer       Hybrid   84512.000000     2
1            Big Data Engineer      On-site   15497.666667     3
2            Big Data Engineer       Remote   66758.333333     3
3                 Data Analyst       Hybrid   39927.500000     6
4                 Data Analyst      On-site   86603.666667    21
5                 Data Analyst       Remote   99319.785714    70
6               Data Architect       Remote  177873.909091    11
7                Data Engineer       Hybrid   65530.416667    12
8                Data Engineer      On-site  101543.740741    27
9                Data Engineer       Remote  122060.795699    93
10        Data Science Manager       Hybrid   74379.500000     2
11        Data Science Manager      On-site  240000.000000     1
12        Data Science Manager       Remote  158772.875000     8
13              Data Scientist       Hybrid   74504.080000    25
14              Data Scientist      On-site   99520.743590    39
15              Data Scientist       Remote  119422.423077    78
16   Machine Learning Engineer       Hybrid   94532.909091    11
17   Machine Learning Engineer      On-site  130172.375000     8
18   Machine Learning Engineer       Remote  100856.590909    22
19  Machine Learning Scientist       Hybrid   95000.000000     3
20  Machine Learning Scientist      On-site  260000.000000     1
21  Machine Learning Scientist       Remote  180575.000000     4
22          Research Scientist       Hybrid   78690.111111     9
23          Research Scientist      On-site  204949.666667     3
24          Research Scientist       Remote  105313.000000     4
25                       other       Hybrid   95244.551724    29
26                       other      On-site  119277.041667    24
27                       other       Remote  134823.651163    86

In [80]:
fig=px.bar(df4remoteandjob,x='job_title',y='mean',color='remote_ratio',barmode='group')
fig.update_layout(yaxis_title= 'Average Salary')
fig.show()

You can see that depending on the job title, remote employees, hybrid employees, and on-site employees will have different wages. While Data Analyst, Data Scientist, Data Engineer have paid more for remote employees, Data Science Manager, Machine Learning Scientist, and Reseach Scientist have given more money to staff.

# Bulding model

In [81]:
df5=df4.copy()

In [82]:
df5

work_year experience_level employment_type                   job_title  \
0        2020        Mid Level       Full time              Data Scientist   
1        2020     Senior Level       Full time  Machine Learning Scientist   
2        2020     Senior Level       Full time           Big Data Engineer   
3        2020        Mid Level       Full time                       other   
4        2020     Senior Level       Full time   Machine Learning Engineer   
..        ...              ...             ...                         ...   
602      2022     Senior Level       Full time               Data Engineer   
603      2022     Senior Level       Full time               Data Engineer   
604      2022     Senior Level       Full time                Data Analyst   
605      2022     Senior Level       Full time                Data Analyst   
606      2022        Mid Level       Full time                       other   

     salary_in_usd employee_residence remote_ratio company_location  \
0            79833                 DE      On-site               DE   
1           260000                 JP      On-site               JP   
2           109024                 GB       Hybrid               GB   
3            20000              other      On-site            other   
4           150000                 US       Hybrid               US   
..             ...                ...          ...              ...   
602         154000                 US       Remote               US   
603         126000                 US       Remote               US   
604         129000                 US      On-site               US   
605         150000                 US       Remote               US   
606         200000                 IN       Remote               US   

    company_size  
0          Large  
1          Small  
2         Medium  
3          Small  
4          Large  
..           ...  
602       Medium  
603       Medium  
604       Medium  
605       Medium  
606        Large  

[605 rows x 9 columns]

In [83]:
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder
from sklearn.pipeline import Pipeline

In [84]:
X=df5.drop('salary_in_usd',axis=1)
y=df5['salary_in_usd']
X_train,X_test,y_train,y_test= train_test_split(X,y,test_size=0.2,random_state=10)
pipeline= Pipeline([('one_hot_cat',OneHotEncoder(handle_unknown='ignore'))])
X_train= pipeline.fit_transform(X_train)


In [85]:
from sklearn.model_selection import ShuffleSplit
from sklearn.model_selection import cross_val_score
cv= ShuffleSplit(n_splits= 5,test_size=0.2, random_state=10)

In [86]:
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import Lasso
from sklearn.linear_model import Ridge
from sklearn.tree import DecisionTreeRegressor
from sklearn.linear_model import LinearRegression
from sklearn.svm import SVC


In [87]:
def find_best_model_using_GridsearchCV(X,y):
    algos= {'linear Regression':{'model':LinearRegression(),'params':{'normalize':[True,False]}},'lasso':{'model':Lasso(),'params':{'alpha':[1,2],'selection':['random','cyclic']}},'decision tree':{'model':DecisionTreeRegressor(),'params':{'criterion':['mse','friedman_mse'],'splitter': ['best','random']}},'ridge':{'model':Ridge(),'params':{'alpha':[1,2],'fit_intercept':[True,False],'normalize':[True,False],'solver':['auto', 'svd', 'cholesky', 'lsqr', 'sparse_cg', 'sag', 'saga', 'lbfgs']}},'svc':{'model':SVC(),'params':{'C':[1.0,2.0],'kernel':['linear', 'poly', 'rbf'],'degree':[1,2,3],'gamma':['scale', 'auto']}}}
    scores=[]
    cv=ShuffleSplit(n_splits=5,test_size=0.2,random_state=0)
    for algo_name, config in algos.items():
        gs= GridSearchCV(config['model'],config['params'],cv=cv,return_train_score=False)
        gs.fit(X,y)
        scores.append({'model':algo_name,'best_score':gs.best_score_,'best_params':gs.best_params_})
    return pd.DataFrame(scores,columns=['model','best_score','best_params'])
     
    

In [88]:
find_best_model_using_GridsearchCV(X_train,y_train)

c:\Users\HP Envy\anaconda3\lib\site-packages\sklearn\linear_model\_base.py:141: FutureWarning:

'normalize' was deprecated in version 1.0 and will be removed in 1.2.
If you wish to scale the data, use Pipeline with a StandardScaler in a preprocessing stage. To reproduce the previous behavior:

from sklearn.pipeline import make_pipeline

model = make_pipeline(StandardScaler(with_mean=False), LinearRegression())

If you wish to pass a sample_weight parameter, you need to pass it as a fit parameter to each step of the pipeline as follows:

kwargs = {s[0] + '__sample_weight': sample_weight for s in model.steps}
model.fit(X, y, **kwargs)



c:\Users\HP Envy\anaconda3\lib\site-packages\sklearn\linear_model\_base.py:141: FutureWarning:

'normalize' was deprecated in version 1.0 and will be removed in 1.2.
If you wish to scale the data, use Pipeline with a StandardScaler in a preprocessing stage. To reproduce the previous behavior:

from sklearn.pipeline import make_pipeline

model = make_pipe

model  best_score  \
0  linear Regression    0.459047   
1              lasso    0.459919   
2      decision tree    0.151940   
3              ridge    0.495524   
4                svc    0.032990   

                                         best_params  
0                                {'normalize': True}  
1                {'alpha': 2, 'selection': 'cyclic'}  
2  {'criterion': 'friedman_mse', 'splitter': 'ran...  
3  {'alpha': 2, 'fit_intercept': False, 'normaliz...  
4  {'C': 1.0, 'degree': 1, 'gamma': 'scale', 'ker...

In [89]:
model=Ridge(alpha='2',fit_intercept=False,normalize=True,solver='auto')

In [90]:
X_test=pipeline.transform(X_test)

In [91]:
model.fit(X_train,y_train)

c:\Users\HP Envy\anaconda3\lib\site-packages\sklearn\linear_model\_base.py:141: FutureWarning:

'normalize' was deprecated in version 1.0 and will be removed in 1.2.
If you wish to scale the data, use Pipeline with a StandardScaler in a preprocessing stage. To reproduce the previous behavior:

from sklearn.pipeline import make_pipeline

model = make_pipeline(StandardScaler(with_mean=False), Ridge())

If you wish to pass a sample_weight parameter, you need to pass it as a fit parameter to each step of the pipeline as follows:

kwargs = {s[0] + '__sample_weight': sample_weight for s in model.steps}
model.fit(X, y, **kwargs)

Set parameter alpha to: original_alpha * n_samples. 



Ridge(alpha='2', fit_intercept=False, normalize=True)

In [92]:
model.score(X_test,y_test)

0.48174250012586894

Predict salary

In [93]:

t=[[2020,'Senior Level','Full time','Data Scientist','US','On-site','GB','Medium']]
t_transform= pipeline.transform(t)
model=Ridge(alpha='2',fit_intercept=False,normalize=True,solver='auto')
model.fit(X_train,y_train)
model.predict(t_transform)


c:\Users\HP Envy\anaconda3\lib\site-packages\sklearn\base.py:450: UserWarning:

X does not have valid feature names, but OneHotEncoder was fitted with feature names

c:\Users\HP Envy\anaconda3\lib\site-packages\sklearn\linear_model\_base.py:141: FutureWarning:

'normalize' was deprecated in version 1.0 and will be removed in 1.2.
If you wish to scale the data, use Pipeline with a StandardScaler in a preprocessing stage. To reproduce the previous behavior:

from sklearn.pipeline import make_pipeline

model = make_pipeline(StandardScaler(with_mean=False), Ridge())

If you wish to pass a sample_weight parameter, you need to pass it as a fit parameter to each step of the pipeline as follows:

kwargs = {s[0] + '__sample_weight': sample_weight for s in model.steps}
model.fit(X, y, **kwargs)

Set parameter alpha to: original_alpha * n_samples. 



array([119086.40573145])